README.rst
============
gocept.month
============

.. image:: https://img.shields.io/pypi/v/gocept.month.svg
    :target: https://pypi.org/project/gocept.month/

.. image:: https://img.shields.io/pypi/pyversions/gocept.month.svg
    :target: https://pypi.org/project/gocept.month/

.. image:: https://travis-ci.com/gocept/gocept.month.svg?branch=master
    :target: https://travis-ci.com/gocept/gocept.month

.. image:: https://coveralls.io/repos/github/gocept/gocept.month/badge.svg?branch=master
    :target: https://coveralls.io/github/gocept/gocept.month?branch=master



A datatype which stores a year and a month.

This package provides the data type ``Month`` (typical usage:
``Month(4, 2003)``), which supports conversion to and from strings, as well as
a zope.schema field and widgets for both zope.formlib and z3c.form.

To use the month field, ``<include package="gocept.month"/>`` and declare a
schema like this:

    >>> import gocept.month
    >>> import zope.interface
    >>> import zope.schema
    >>> class IContract(zope.interface.Interface):
    ...     title = zope.schema.TextLine(title=u"Contract title")
    ...     start = gocept.month.MonthField(title=u"Starting date")
    ...     end = gocept.month.MonthField(title=u"Ending date")

To use the form widgets, you need to require the ``form`` setuptools extra
(i.e. ``gocept.month[form]``), and ``<include
package="gocept.month.browser"/>`` in your ZCML.


CHANGES.rst
CHANGES
=======

3.1 (unreleased)
----------------

- Drop support for Python 3.6.

- Add support for Python 3.11.


3.0 (2021-08-27)
----------------

Backwards-incompatible changes
++++++++++++++++++++++++++++++

- Re-licence from ZPL to MIT licence.

- Drop support for Python 2 and 3.6.

Features
++++++++

- Add support for Python 3.10 (as of rc.1).


2.2 (2021-04-20)
----------------

Features
++++++++

- Add support for Python 3.7, 3.8 and 3.9.

Bug fixes
+++++++++

- Add missing security declaration needed for ``.browser.widget.MonthWidget``,
  the month widget for ``z3c.form``.

Other CHANGES
+++++++++++++

- Migrate to Github.

- Use Github actions as CI.

- Drop support for Python 3.4 and 3.5.


2.1 (2017-04-13)
----------------

- Officially support Python 3.6.

- Fix `configure.zcml` to be compatible with Python 3.


2.0 (2016-04-28)
----------------

- Made the package compatible with Python 2.7 and Python 3 at the same time.

- Backwards-incompatible change towards stricter comparison behaviour: While
  arbitrary objects (except month strings) used to be considered less than any
  ``Month`` instance, such a comparison now raises a ``TypeError``.

- Removed deprecated ``Month.isBetween`` method.


1.5 (2016-04-27)
----------------

- Support Python 2.7 and PyPy only.

- Add tox as testrunner.


1.4 (2016-02-07)
----------------

- Add a ``__len__`` method to `MonthInterval`.

- Require `z3c.form >= 2.6` to get rid of some odd test dependencies.


1.3.2 (2015-10-13)
------------------

- Fix `configure.zcml` to include needed `meta.zcml` files.


1.3.1 (2015-08-05)
------------------

- Migrate repository to https://bitbucket.org/gocept/gocept.month


1.3 (2014-07-16)
----------------

- Add placeholder for month widget in z3c forms describing the required month
  format.


1.2 (2013-02-18)
----------------

- Split off widgets into setuptools extra ``form``, and don't include the
  ``browser`` ZCML by ourselves, so clients can use only the domain part of
  this package without the UI parts.

- Added ``firstOfMonth`` method to ``Month``.

- Using `zope.formlib >= 4.0` instead of `zope.app.form`.

- Updated to ZTK 1.1.5.


1.1 (2012-02-09)
----------------

- Add MonthInterval.forYear().
- Declared ``fromString`` method as a class method in interface.


1.0.2 (2011-09-01)
------------------

- Added ``__contains__`` method on ``Month``.


1.0.1 (2011-04-26)
------------------

- Declare dependencies that some upstream packages need but don't declare.

- Use stdlib's doctest module, not zope.testing's which is deprecated.


1.0 (2009-10-05)
----------------

- Initial public release.


src/gocept/month/tests/test_month.py
from .. import Month
import gocept.month
import pytest


@pytest.fixture
def m():
    return Month(11, 2005)


@pytest.fixture(scope='module')
def zca(request):
    """Load ZCML on session module."""
    import plone.testing.zca
    layer = plone.testing.zca.ZCMLSandbox(
        name='MonthZCML', filename='configure.zcml',
        module=__name__, package=gocept.month)
    layer.setUp()
    yield layer
    layer.tearDown()


class FakeMonth:
    month = 11
    year = 2005


fakemonth = FakeMonth()


def test_eq(m):
    assert m != None  # noqa
    assert m != fakemonth
    assert m == Month(11, 2005)
    assert m != Month(10, 2005)
    assert m != Month(11, 2004)


def test_lt(m):
    assert m < Month(11, 2006)
    assert m < Month(10, 2006)
    assert m < Month(12, 2005)
    assert not (m < Month(11, 2005))
    assert not (m < Month(10, 2005))
    assert not (m < Month(11, 2004))
    assert not (m < Month(12, 2004))
    with pytest.raises(TypeError):
        m < None
    with pytest.raises(TypeError):
        m < fakemonth


def test_gt(m):
    assert not (m > Month(11, 2006))
    assert not (m > Month(10, 2006))
    assert not (m > Month(12, 2005))
    assert not (m > Month(11, 2005))
    assert m > Month(10, 2005)
    assert m > Month(11, 2004)
    assert m > Month(12, 2004)
    with pytest.raises(TypeError):
        m > None
    with pytest.raises(TypeError):
        m > fakemonth


def test_le(m):
    assert m <= Month(11, 2006)
    assert m <= Month(10, 2006)
    assert m <= Month(12, 2005)
    assert m <= Month(11, 2005)
    assert not (m <= Month(10, 2005))
    assert not (m <= Month(11, 2004))
    assert not (m <= Month(12, 2004))
    with pytest.raises(TypeError):
        m <= None
    with pytest.raises(TypeError):
        m <= fakemonth


def test_ge(m):
    assert not (m >= Month(11, 2006))
    assert not (m >= Month(10, 2006))
    assert not (m >= Month(12, 2005))
    assert m >= Month(11, 2005)
    assert m >= Month(10, 2005)
    assert m >= Month(11, 2004)
    assert m >= Month(12, 2004)
    with pytest.raises(TypeError):
        m >= None
    with pytest.raises(TypeError):
        m >= fakemonth


def test_eq_str(m, zca):
    assert m == '11/2005'
    assert m != '10/2005'
    assert m != '11/2004'


def test_lt_str(m, zca):
    assert m < '11/2006'
    assert m < '10/2006'
    assert m < '12/2005'
    assert not (m < '11/2005')
    assert not (m < '10/2005')
    assert not (m < '11/2004')
    assert not (m < '12/2004')


def test_gt_str(m, zca):
    assert not (m > '11/2006')
    assert not (m > '10/2006')
    assert not (m > '12/2005')
    assert not (m > '11/2005')
    assert m > '10/2005'
    assert m > '11/2004'
    assert m > '12/2004'


def test_le_str(m, zca):
    assert m <= '11/2006'
    assert m <= '10/2006'
    assert m <= '12/2005'
    assert m <= '11/2005'
    assert not (m <= '10/2005')
    assert not (m <= '11/2004')
    assert not (m <= '12/2004')


def test_ge_str(m, zca):
    assert not (m >= '11/2006')
    assert not (m >= '10/2006')
    assert not (m >= '12/2005')
    assert m >= '11/2005'
    assert m >= '10/2005'
    assert m >= '11/2004'
    assert m >= '12/2004'


def test_eq_str_left(m, zca):
    assert '11/2005' == m
    assert '10/2005' != m
    assert '11/2004' != m


def test_lt_str_left(m, zca):
    assert '11/2006' > m
    assert '10/2006' > m
    assert '12/2005' > m
    assert not ('11/2005' > m)
    assert not ('10/2005' > m)
    assert not ('11/2004' > m)
    assert not ('12/2004' > m)


def test_gt_str_left(m, zca):
    assert not ('11/2006' < m)
    assert not ('10/2006' < m)
    assert not ('12/2005' < m)
    assert not ('11/2005' < m)
    assert '10/2005' < m
    assert '11/2004' < m
    assert '12/2004' < m


def test_le_str_left(m, zca):
    assert '11/2006' >= m
    assert '10/2006' >= m
    assert '12/2005' >= m
    assert '11/2005' >= m
    assert not ('10/2005' >= m)
    assert not ('11/2004' >= m)
    assert not ('12/2004' >= m)


def test_ge_str_left(m, zca):
    assert not ('11/2006' <= m)
    assert not ('10/2006' <= m)
    assert not ('12/2005' <= m)
    assert '11/2005' <= m
    assert '10/2005' <= m
    assert '11/2004' <= m
    assert '12/2004' <= m


src/gocept/month/_month.py
from gocept.month.interfaces import IMonth
import datetime
import functools
import gocept.month
import re
import zope.interface


@zope.interface.implementer(IMonth)
@functools.total_ordering
class Month:
    """A datatype which stores a year and a month.

    >>> from zope.interface.verify import verifyObject
    >>> verifyObject(IMonth, Month(11, 1977))
    True
    """

    month = property(lambda self: self.__month)
    year = property(lambda self: self.__year)

    def __init__(self, month, year):
        """Constructor

        >>> Month(13, 2005)
        Traceback (most recent call last):
        ...
        ValueError: Month must be between 1 and 12.
        >>> Month(0, 2005)
        Traceback (most recent call last):
        ...
        ValueError: Month must be between 1 and 12.
        >>> Month(11, 2005)
        Month 11/2005
        >>> Month(11,5)
        Month 11/5
        >>> Month(11,99)
        Month 11/99
        >>> Month(11,-20)
        Traceback (most recent call last):
        ...
        ValueError: Year must be at least 1.
        """
        month = int(month)
        if not (1 <= month <= 12):
            raise ValueError('Month must be between 1 and 12.')
        self.__month = month

        year = int(year)
        if year <= 0:
            raise ValueError("Year must be at least 1.")
        self.__year = year

    def __repr__(self):
        return f"Month {self.month}/{self.year}"

    def __eq__(self, other):
        """Compare for equality. Not implementing IMonth means inequality."""
        try:
            other = IMonth(other)
        except TypeError:
            return False
        else:
            return (self.year, self.month) == (other.year, other.month)

    def __gt__(self, other):
        """Compare for strict ordering (greater than other).

        If other is not adaptable to IMonth it is considered a TypeError.

        """
        other = IMonth(other)
        return (self.year, self.month) > (other.year, other.month)

    date_regex = re.compile(r"^([0-9]{1,2})[,./-]?([0-9]{2}|[0-9]{4})$")

    @classmethod
    def current(cls):
        """Return a month instance for the current month.

        >>> Month.current() # doctest: +ELLIPSIS
        Month .../...

        """
        now = datetime.date.today()
        return Month(now.month, now.year)

    @classmethod
    def fromString(cls, string):
        """Get instance from a string. The Month must come first in string.

        >>> Month.fromString('11/2005')
        Month 11/2005
        >>> Month.fromString('11.2005')
        Month 11/2005
        >>> Month.fromString('11-2005')
        Month 11/2005
        >>> Month.fromString('11.2005')
        Month 11/2005
        >>> Month.fromString('1.2005')
        Month 1/2005
        >>> Month.fromString('01.2005')
        Month 1/2005
        >>> Month.fromString('11.05')
        Month 11/2005
        >>> Month.fromString('11/05')
        Month 11/2005
        >>> Month.fromString('1105')
        Month 11/2005
        >>> Month.fromString('112005')
        Month 11/2005
        >>> Month.fromString('105')
        Month 1/2005
        >>> Month.fromString('12005')
        Month 1/2005
        >>> Month.fromString('0105')
        Month 1/2005
        >>> Month.fromString('0501')
        Month 5/2001
        >>> Month.fromString('012005')
        Month 1/2005

        Empty strings result in None instead of a Month instance

        >>> Month.fromString('') is None
        True

        Format Errors result in a ValueError

        >>> Month.fromString('2005-11')
        Traceback (most recent call last):
        ...
        ValueError: Date must be MM/YYYY.
        """
        if string == '':
            return None
        result = cls.date_regex.match(string)
        if result is None:
            raise ValueError('Date must be MM/YYYY.')
        month, year = result.groups()
        if len(year) == 2:
            year = "20%s" % year
        return Month(int(month), int(year))

    def firstOfMonth(self):
        """Get the datetime.date which represents the first day of the month.

        >>> Month(11, 2005).firstOfMonth()
        datetime.date(2005, 11, 1)
        >>> Month(1, 2000).firstOfMonth()
        datetime.date(2000, 1, 1)
        """
        return datetime.date(self.year, self.month, 1)

    def lastOfMonth(self):
        """Get the datetime.date which represents the last day of the month.

        >>> Month(11, 2005).lastOfMonth()
        datetime.date(2005, 11, 30)
        >>> Month(12, 2005).lastOfMonth()
        datetime.date(2005, 12, 31)
        >>> Month(2, 2008).lastOfMonth()
        datetime.date(2008, 2, 29)
        """
        next_month = self + 1
        return next_month.firstOfMonth() - datetime.timedelta(days=1)

    def __sub__(self, other):
        """Substract from this month.

        Given an `int` a previous month is computed:

        >>> Month(11, 2005) - 10
        Month 1/2005
        >>> Month(1, 2005) - 13
        Month 12/2003
        >>> Month(1, 2005) - 5
        Month 8/2004

        Given an `IMonth` an interval is computed:

        >>> Month(11, 2005) - Month(1, 2005)
        <MonthInterval from Month 11/2005 to Month 1/2005>
        >>> Month(1, 2005) - Month(11, 2005)
        <MonthInterval from Month 1/2005 to Month 11/2005>

        Trying to subtract other types raises an error:
        >>> Month(11, 2005) - '1'
        Traceback (most recent call last):
        TypeError: Can't subtract <... 'str'> from month.

        """
        if isinstance(other, int):
            months = other
            year = self.year - (months // 12)
            month = self.month - months % 12
            if month <= 0:
                year -= 1
                month += 12
            return Month(month, year)
        elif IMonth.providedBy(other):
            return gocept.month.MonthInterval(self, other)
        raise TypeError("Can't subtract %r from month." % type(other))

    def __add__(self, months):
        """Add months and return a new IMonth.

        >>> m1 = Month(12,2005)
        >>> m2 = Month(1,2005)
        >>> m1 + 1
        Month 1/2006
        >>> m2 + 1
        Month 2/2005
        >>> m2 + 13
        Month 2/2006
        """
        return self - (-months)

    def __str__(self):
        """Returns a string representation.

        >>> str(Month(10,2000))
        '10/2000'
        >>> str(Month(1,2005))
        '01/2005'
        """
        return "%02i/%s" % (self.month, self.year)

    def __hash__(self):
        """Returns the hash.

        Make sure Month is immutable:

        >>> Month(5,2003).year = 2005
        Traceback (most recent call last):
        ...
        AttributeError: ...
        >>> Month(5,2003).month = 4
        Traceback (most recent call last):
        ...
        AttributeError: ...

        Check the hash

        >>> hash(Month(10,2000))
        102000
        >>> hash(Month(1,2005))
        12005
        >>>
        """
        return int(f"{self.month}{self.year}")

    def __iter__(self):
        """Returns an iterator over the days of the month.

        Represents each day as a datetime.date.

        >>> days = iter(Month(5, 2009))
        >>> next(days)
        datetime.date(2009, 5, 1)
        >>> next(days)
        datetime.date(2009, 5, 2)
        >>> for day in days: pass
        >>> day
        datetime.date(2009, 5, 31)

        """
        return (datetime.date(self.year, self.month, day)
                for day in range(1, self.lastOfMonth().day + 1))

    def __contains__(self, date):
        """Returns whether the `date` is in the month.

        >>> import datetime
        >>> datetime.date(2009, 4, 30) in Month(5, 2009)
        False
        >>> datetime.date(2009, 5, 1) in Month(5, 2009)
        True
        >>> datetime.date(2009, 5, 2) in Month(5, 2009)
        True
        >>> datetime.date(2009, 5, 31) in Month(5, 2009)
        True
        >>> datetime.date(2009, 6, 1) in Month(5, 2009)
        False
        >>> datetime.date(2012, 2, 29) in Month(2, 2012)
        True
        >>> datetime.datetime(2012, 2, 29, 15, 7, 34) in Month(2, 2012)
        True
        >>> object() in Month(2, 2012)
        False

        """
        if not isinstance(date, datetime.date):
            return False
        return date.year == self.year and date.month == self.month


src/gocept/month/interfaces.py
from zope.interface import Attribute
from zope.interface import Interface
from zope.schema.interfaces import IField
from zope.schema.interfaces import IMinMax
import sys


class IMonthClass(Interface):
    """Static methods of Month"""

    def current():
        """Return a month instance for the current month."""

    def fromString(string):
        """Get instance from a string.

        Raises ValueError if not convertable to month.
        """


class IMonth(Interface):
    """A datatype which stores a year and a month."""

    month = Attribute(
        'Month part of the date. Must be an int between 1 and 12')
    year = Attribute('Four digit year.')

    def __eq__(other):
        """Compare for equality. Not implementing IMonth means inequality."""

    if sys.version_info.major >= 3:
        def __ne__(other):
            """Compare for inequality.

            Not implementing IMonth means inequality.

            """

    def __gt__(other):
        """Compare for strict ordering (greater than other).

        If other is not adaptable to IMonth it is considered a TypeError.

        """

    def __lt__(other):
        """Compare for strict ordering (less than other).

        If other is not adaptable to IMonth it is considered a TypeError.

        """

    def __ge__(other):
        """Compare for ordering (greater than or equal to other).

        If other is not adaptable to IMonth it is considered a TypeError.

        """

    def __le__(other):
        """Compare for ordering (less than or equal to other).

        If other is not adaptable to IMonth it is considered a TypeError.

        """

    def firstOfMonth():
        """Get the date object which represents the first day of the month."""

    def lastOfMonth():
        """Get the date object which represents the last day of the month."""

    def __sub__(other):
        """Substract from this month.

        Given an `int` compute a previous month.
        Given an `IMonth` compute the interval.

        """

    def __add__(months):
        """Add a given number of months."""

    def __str__():
        """Return a string representation."""

    def __hash__():
        """Return the hash."""

    def __iter__():
        """Return an iterator over the days of the month.

        Represents each day as a datetime.date.
        """

    def __contains__(date):
        """Return whether the `date` is in the month."""


class IMonthInterval(Interface):
    """represents an interval between two months"""

    def intersects(other):
        """check if the given interval has an intersection"""

    def normalized():
        """Create a normalized form of this interval.

        The following conditions apply to normalized intervals:

        - start <= end

        """

    def __iter__():
        """Iterate over all months that are in the interval,

        Including start and end.
        """

    def __contains__(month):
        """Return true if the given month is within the interval."""

    def __len__():
        """Return the number of months in the normalized interval.

        An interval with the same start and end has the length of 1!
        """

    def forYear(year):
        """(classmethod) Returns an interval of months for the given year."""


class IMonthField(IMinMax, IField):
    """Field containing a Month."""


class IDate(Interface):
    """A date containing of month and year.

    This is provided by `datetime.date` and can be adapted to `IMonth`.
    """

    month = Attribute('month')
    year = Attribute('year')


src/gocept/month/browser/widget.py
# coding: utf-8
from gocept.month import Month
from zope.formlib.textwidgets import escape, TextWidget
from zope.formlib.widget import DisplayWidget, renderElement
from zope.formlib.interfaces import (
    IDisplayWidget, IInputWidget, ConversionError)
import gocept.month.field
import z3c.form.browser.text
import z3c.form.error
import z3c.form.interfaces
import z3c.form.widget
import zope.component
import zope.i18nmessageid
import zope.interface
import zope.schema.interfaces

_ = zope.i18nmessageid.MessageFactory("gocept")


class MonthConversionError(ConversionError):
    __doc__ = _(u"Format muss MM/JJJJ sein (z.B. 11/2005 für November 2005).")

    def doc(self):
        return self.__class__.__doc__


@zope.interface.implementer(IDisplayWidget)
class MonthDisplayWidget(DisplayWidget):
    """Widget displaying the contents of an IMonthField."""

    cssClass = "month"

    def __call__(self):
        if self._renderedValueSet():
            content = self._data
        else:
            content = self.context.default
        if content == self.context.missing_value:
            return ""
        content = str(content)
        return renderElement("span",
                             contents=escape(content),
                             cssClass=self.cssClass)


@zope.interface.implementer(IInputWidget)
class MonthEditWidget(TextWidget):
    """Widget displaying the contents of an IMonthField. """

    def _toFieldValue(self, input):
        try:
            return Month.fromString(input)
        except ValueError, error:
            raise MonthConversionError(error)


class MonthWidget(z3c.form.browser.text.TextWidget):
    """Widget displaying the contents of an IMonthField for z3c.form. """


@zope.component.adapter(zope.schema.interfaces.IField,
                        z3c.form.interfaces.IFormLayer)
@zope.interface.implementer(z3c.form.interfaces.IFieldWidget)
def MonthFieldWidget(field, request):
    return z3c.form.widget.FieldWidget(field, MonthWidget(request))


class MonthErrorViewSnippet(z3c.form.error.ErrorViewSnippet):

    zope.component.adapts(
        ValueError, None, None, gocept.month.field.MonthField, None,
        None)

    def createMessage(self):
        return MonthConversionError.__doc__


src/gocept/month/interval.py
from gocept.month import IMonthInterval, IMonth, Month
import gocept.month
import zope.interface


@zope.interface.implementer(IMonthInterval)
class MonthInterval(object):
    """Represents a month interval.

    >>> from zope.interface.verify import verifyClass
    >>> verifyClass(IMonthInterval, MonthInterval)
    True
    """

    start = None
    end = None

    def __init__(self, start, end):
        """Create an interval between start and end.

        Start and end must be months:

        >>> from gocept.month import Month
        >>> MonthInterval(1,2)
        Traceback (most recent call last):
        TypeError: start 1 is not a month.
        >>> MonthInterval(Month(1, 2012), Month(3, 2012))
        <MonthInterval from Month 1/2012 to Month 3/2012>

        Alternatively None can be given to indicate open intervals:

        >>> MonthInterval(None, Month(3, 2012))
        <MonthInterval from Month 1/1 to Month 3/2012>
        >>> MonthInterval(Month(2, 2005), None)
        <MonthInterval from Month 2/2005 to Month 12/9999>

        Reverse intervals are also supported:

        >>> MonthInterval(Month(3, 2012), Month(1, 2012))
        <MonthInterval from Month 3/2012 to Month 1/2012>

        """
        if start is None:
            start = gocept.month.Month(1, 1)
        if end is None:
            end = gocept.month.Month(12, 9999)

        if not IMonth.providedBy(start):
            raise TypeError("start %r is not a month." % start)
        if not IMonth.providedBy(end):
            raise TypeError("end %r is not a month." % end)

        self.start = start
        self.end = end

    def __repr__(self):
        return '<MonthInterval from %r to %r>' % (self.start, self.end)

    def __iter__(self):
        """Iterate over all months in the interval, including start and end.

        >>> from gocept.month import Month
        >>> for m in MonthInterval(Month(1, 2005), Month(12, 2005)):
        ...     print m
        01/2005
        02/2005
        03/2005
        04/2005
        05/2005
        06/2005
        07/2005
        08/2005
        09/2005
        10/2005
        11/2005
        12/2005

        >>> for m in MonthInterval(Month(11, 2005), Month(9, 2005)):
        ...     print m
        11/2005
        10/2005
        09/2005

        >>> list(MonthInterval(Month(11, 2005), Month(11, 2005)))
        [Month 11/2005]

        >>> open_interval = MonthInterval(Month(11, 2005), None)
        >>> i = iter(open_interval)
        >>> print i.next()
        11/2005
        >>> print i.next()
        12/2005

        >>> open_interval = MonthInterval(None, Month(11, 2005))
        >>> i = iter(open_interval)
        >>> print i.next()
        01/1
        >>> print i.next()
        02/1

        """
        if self.start <= self.end:
            op = +1
        else:
            op = -1

        month = self.start
        while True:
            yield month
            if month == self.end:
                return
            month += op

    def __len__(self):
        """Return the number of months in the normalized interval.

        >>> from gocept.month import Month
        >>> len(MonthInterval(Month(1, 2016), Month(1, 2016)))
        1
        >>> len(MonthInterval(Month(1, 2016), Month(2, 2016)))
        2
        >>> len(MonthInterval(Month(1, 2016), Month(12, 2016)))
        12
        >>> len(MonthInterval(Month(1, 2016), Month(1, 2017)))
        13
        >>> len(MonthInterval(Month(2, 2016), Month(1, 2016)))
        2
        """
        i = self.normalized()
        years = i.end.year - i.start.year
        months = i.end.month - i.start.month
        return months + years * 12 + 1

    def intersects(self, other):
        """Check whether this interval shares months with another interval.

        >>> from gocept.month import Month
        >>> i = MonthInterval(Month(12, 4), Month(6, 5))
        >>> i.intersects(4)
        Traceback (most recent call last):
        ...
        TypeError: Not an interval: 4
        >>> i.intersects(MonthInterval(Month(5, 5), Month(7, 5)))
        True
        >>> i.intersects(MonthInterval(Month(6, 5), Month(8, 5)))
        True
        >>> i.intersects(MonthInterval(Month(7, 5), Month(9, 5)))
        False
        >>> i.intersects(MonthInterval(Month(11, 4), Month(1, 5)))
        True
        >>> i.intersects(MonthInterval(Month(10, 4), Month(12, 4)))
        True
        >>> i.intersects(MonthInterval(Month(9, 4), Month(11, 4)))
        False
        >>> i.intersects(MonthInterval(Month(2, 5), Month(04, 5)))
        True
        >>> i.intersects(i)
        True
        >>> i.intersects(MonthInterval(Month(9, 4), Month(3, 5)))
        True
        >>> i.intersects(MonthInterval(Month(3, 5), Month(9, 5)))
        True
        >>> MonthInterval(Month(9, 4), Month(9, 5)).intersects(i)
        True
        >>> i.intersects(MonthInterval(Month(9, 4), Month(9, 5)))
        True
        >>> i.intersects(MonthInterval(None, None))
        True
        >>> MonthInterval(None, None).intersects(MonthInterval(None, None))
        True
        >>> i.intersects(MonthInterval(None, Month(1, 5)))
        True
        >>> MonthInterval(None, Month(12, 4)).intersects(i)
        True
        >>> i.intersects(MonthInterval(None, Month(12, 4)))
        True
        >>> i.intersects(MonthInterval(None, Month(11, 4)))
        False
        >>> i.intersects(MonthInterval(Month(5, 5), None))
        True
        >>> i.intersects(MonthInterval(Month(6, 5), None))
        True
        >>> MonthInterval(Month(7, 5), None).intersects(i)
        False
        >>> i.intersects(MonthInterval(Month(7, 5), None))
        False

        Intersection also works with reverse intervals:

        >>> i.intersects(MonthInterval(Month(7, 5), Month(3, 5)))
        True

        """
        if not IMonthInterval.providedBy(other):
            raise TypeError("Not an interval: %r" % other)
        a = self.normalized()
        b = other.normalized()
        if b.end < a.start:
            return False
        if a.end < b.start:
            return False
        return True

    def __contains__(self, month):
        """Return `True` if the given month is within the interval.

        >>> from gocept.month import Month
        >>> interval = MonthInterval(Month(1, 2001), Month(5, 2001))
        >>> Month(1, 2001) in interval
        True
        >>> Month(5, 2001) in interval
        True
        >>> Month(3, 2001) in interval
        True
        >>> Month(12, 2000) in interval
        False
        >>> Month(6, 2001) in interval
        False
        >>> Month(10, 1999) in interval
        False
        >>> Month(10, 2007) in interval
        False

        Also works for reverse intervals:

        >>> Month(5, 2003) in MonthInterval(Month(10, 2003), Month(3, 2003))
        True

        """
        a = self.normalized()
        if a.start <= month <= a.end:
            return True
        return False

    def normalized(self):
        """Create a normalized form of this interval.

        >>> from gocept.month import Month
        >>> interval = MonthInterval(Month(5, 2001), Month(1, 2001))
        >>> interval.normalized()
        <MonthInterval from Month 1/2001 to Month 5/2001>
        >>> interval.normalized().normalized()
        <MonthInterval from Month 1/2001 to Month 5/2001>

        """
        if self.start <= self.end:
            start, end = self.start, self.end
        else:
            start, end = self.end, self.start
        return MonthInterval(start, end)

    @classmethod
    def forYear(cls, year):
        """Return an interval of months for the given year.

        >>> MonthInterval.forYear(2001)
        <MonthInterval from Month 1/2001 to Month 12/2001>
        """
        return cls(Month(1, year), Month(12, year))
